Fixes #17880 - Follow same pattern for DNS CNAME methods signature

guimaluf opened the pull request:
All DNS methods uses fqdn parameters as the target of the action and the    second argument as the attribute that will be set to that target

Since CNAME methods does the opposite, this commit aim to fix that

mmoll commented:
[test]

ekohl commented:
The annoying thing is that plugins can have implemented this feature using this API. Now the number of DNS plugins is limited, but we've already shipped this API in 1.13. I'm still working on implementing it for PowerDNS so I hadn't come across this yet.

This includes https://github.com/theforeman/smart-proxy/pull/489 and I see some other fixes in the imports. Let's first get those merged and then look at this again.

dmitri-d commented:
@guimaluf: could you rebase this PR and fix the conflicts please?

guimaluf committed:
Fixes #17880 - Keep pattern for DNS CNAME methods signature

All DNS methods uses fqdn parameters as the target of the action and
the second argument as the attribute that will be set to that
target

Since CNAME methods does the opposite, this commit aim to
fix that

guimaluf commented on test/dns/dns_api_test.rb:
```diff
@@ -149,10 +149,10 @@ def test_create_aaaa_record_is_shortened
 
   def test_create_cname_record
-    post '/', :fqdn => 'test.com', :value => 'test1.com', :type => 'CNAME'
+    post '/', :fqdn => 'test.com', :value => 'alias.com', :type => 'CNAME'
```
This is the main important thing about this PR.
It'll change the signature of cname creation API.
I don't know if we should point out deprecation warning or keep a backwords compatibility

ekohl commented on modules/dns_common/dns_common.rb:
```diff
@@ -170,10 +170,10 @@ def aaaa_record_conflicts(fqdn, ip)
-    def cname_record_conflicts(fqdn, target)
-      current = resolver.getresources(fqdn, Resolv::DNS::Resource::IN::CNAME)
+    def cname_record_conflicts(fqdn, host_alias)
+      current = resolver.getresources(host_alias, Resolv::DNS::Resource::IN::CNAME)
```
I don't think this is correct. If I interpret it correctly then ```fqdn``` is the name and ```target``` the value. https://github.com/theforeman/smart-proxy/blob/47732b9ef1375e7bc88b7056f497080477ac2e31/modules/dns/dns_api.rb#L30 passes ```fqdn```, ```value```. Then https://github.com/theforeman/smart-proxy/blob/47732b9ef1375e7bc88b7056f497080477ac2e31/modules/dns_common/dns_common.rb#L45-L46 does the same.

guimaluf replied:
It works already, but from my point of view this signature is incorrect.

`server.create_cname_record(fqdn, value)`
Actual CNAME implementations `fqdn` refers to host alias, while `value` refers to the true `fqdn`(`target`).

It makes more sense to use `fqdn=host.example.com, value=alias.example.com`, like IP attribution, `fqdn=host.example.com, value=10.0.0.1`.

Why CNAME is the only that would does the opposite?

ekohl replied:
To me the DNS API is not DNS API but a Host to DNS translation API. That makes it inconsistent and hard to say which is 'correct' because it is a custom format.

The way I'd want it is that fqdn refers to what you query the server with and value what the server answers with. Since you query alias.example.com, that should be the value. In short: ```dig +short $type $fqdn``` should result in ```$value```. However, for PTRs this is already reversed. The way Foreman defines it this is that fqdn is the Host object and value is how it should point to the host object.

The current implementation already aligns with A/AAAA records. I'd say that PTR should be aligned with A/AAAA/CNAME rather than align CNAME with PTR. However, this is all breaking the API which we just can't do since this is all in released versions.

I'd propose to internally standardize the naming:
* ```fqdn``` is what is defined in the API and should only be in dns_api.rb
* ```value``` is what is defined in the API and should only be in dns_api.rb
* ```name``` is the DNS name you query for
* ```content``` is the result your DNS server should return

We should leave the current argument orders since this is the plugin API and breaking them is something we should avoid.

guimaluf replied:
I get your point. Thanks for clarifying. 
So, should we standardize this or leave as it is?

dmitri-d replied:
I agree with @ekohl, mostly. As this is internal api, deprecating a call and introducing a new one *is* an option: this way provider developers have about three months to catch up. Having said this, I'd rate good method parameter names higher than consistent parameter order, esp. when we only have a couple of parameters in these methods.

ekohl reviewed (changes requested):
I'm not sure this is even a bug but suspect you're interpreting the API incorrectly.

When you do a POST with fqdn=alias.example.com value=host.example.com then it should already work correctly. I believe I've verified this with bind and powerdns on a system level, using dig to verify.

dmitri-d commented:
As the discussion seems to have stopped, I'm going to assume that we have settled on not making this change. @guimaluf: Please re-open the PR if you'd like to see it merged.

ekohl commented:
I do intend to submit a PR that makes the parameter names consistent like I suggested above.

ekohl commented:
Created https://github.com/theforeman/smart-proxy/pull/507 to fix this.
